### ISSUE TITLE
Excessive memory consumption on project search with large files present

### ISSUE THREAD

**Issue by notpeter** [zed-industries/zed#20970]
tldr: Project search in directories with big files will cause Zed to crash after consuming all your memory.

Play stupid games, win stupid prizes.

Steps to reproduce:
1. Run:
```
mkdir -p /tmp/protein
cd $_
wget https://aiweb.cs.washington.edu/research/projects/xmltk/xmldata/data/pir/psd7003.xml.gz
gunzip psd7003.xml.gz
rm psd7003.xml.gz
for i in {1..30}; do cp psd7003.xml psd-$i.xml; done
zed .
```
2. Quit/Save work in any apps you care about, because bad things can happen under memory pressure.
3. `cmd-shift-f` and search for `<`

This will rapidly consume memory and on my machine crashes after ~5secs.
On my machine if you use 20 files (14GB) instead of 30 (21GB) it takes a while but will finish.

Watch out though, I later ran it with 50 and caused a kernel panic.

### Environment

Zed: v0.163.0 (Zed Nightly 3c57a4071cd3e3400a1e0b3329fccb0f0477e1ea)
OS: macOS 14.7.0
Memory: 64 GiB
Architecture: aarch64

**Comment by SomeoneToIgnore** [comment]
This seems to be a combination of XML language server issues (leaks reported) + Zed starting a language server for each excerpt (also reported).

So technically a duplicate, but a good repro case for both issues.

**Comment by notpeter** [comment]
I don't think this is a language server thing.
It happens when searching in Plain Text too.
Amazingly Zed beach-balled, ballooned to over 100GB of ram, finished the search, rendered the multi-buffer and dropped to 8GB of ram. When I closed the multi-buffer it dropped to 450MB.
Bravo Zed. Truly stunning.

While this is a bit exaggerated (40GB of XML == 120GB of memory) it's not that crazy for folks to have a directory with data files in a folder that they are also editing in Zed.  In practice it means if you project search for the wrong needle with a large datasets lying around you can KP your mac (I did).

```
mkdir /tmp/uspto-txt
cd /tmp/uspto-txt
wget https://bulkdata.uspto.gov/data/patent/grant/redbook/fulltext/2024/ipg241119.zip
unzip ipg241119.zip
rm ipg241119.zip
for i in {1..50}; do cp ipg241119.xml ipg241119-$i.txt; done
rm ipg241119.xml
zed .
```

Seriously. Save your work before trying this.
And when macOS suggests you kill Zed, do it.
Failure to so may result in system freezing.

<img width="788" alt="Screenshot 2024-11-21 at 1 23 25" src="https://github.com/user-attachments/assets/20cd46b5-6ace-4ecf-ab06-194043a9f201">

**Comment by RemcoSmitsDev** [comment]
This might be related: https://github.com/zed-industries/zed/issues/20436

**Comment by ElanUtta** [comment]
@marcospb19 Is there a view of resolving this problem? There is something I can help with?

**Comment by notpeter** [comment]
I can still reproduce this.
I updated the description to reflect the current behavior (crash instead of macOS OOM) and to use different files as the example USPTO zip file is now 404s.

**Comment by drahnr** [comment]
I can reproduce the out of memory condition at will using the search and replace functionality on medium sized codebase.

`Zed 0.195.5 72015b2f6636e154680a570f402736b6c928a14f  – /home/bernhard/.local/zed.app/libexec/zed-editor`

EDIT: Enabling https://github.com/0xMiden/miden-zed-extension and opening the `next` branch of  https://github.com/0xMiden/miden-node consistently triggers it.

**Comment by sitar777** [comment]
Not only XML but reproduces with Python language server too
